File: sdks/python/kavachos/delegation.py
"""Delegation chain management."""

from __future__ import annotations

from typing import List
from urllib.parse import quote

from kavachos._http import AsyncTransport, SyncTransport
from kavachos.types import DelegateInput, DelegationChain, Permission


class AsyncDelegationResource:
    """Async delegation chain operations."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._http = transport

    async def create(self, input: DelegateInput) -> DelegationChain:
        """Create a delegation from one agent to another.

        The delegated agent receives a subset of the granting agent's
        permissions, subject to the depth limit.

        Args:
            input: Delegation parameters including agents, permissions, and
                expiry.

        Returns:
            The newly created :class:`~kavachos.types.DelegationChain`.
        """
        data = await self._http.request("POST", "/delegations", json=input.to_dict())
        return DelegationChain.from_dict(data)

    async def list_chains(self, agent_id: str) -> List[DelegationChain]:
        """List all delegation chains where the agent is a participant.

        Args:
            agent_id: The agent's unique identifier.

        Returns:
            List of :class:`~kavachos.types.DelegationChain` objects.
        """
        data = await self._http.request("GET", f"/delegations/{quote(agent_id)}")
        return [DelegationChain.from_dict(d) for d in data]

    async def revoke(self, delegation_id: str) -> None:
        """Revoke a specific delegation chain.

        Args:
            delegation_id: The delegation's unique identifier.
        """
        await self._http.request("DELETE", f"/delegations/{quote(delegation_id)}")

    async def get_effective_permissions(self, agent_id: str) -> List[Permission]:
        """Get the effective permissions for an agent, merging own and delegated.

        Args:
            agent_id: The agent's unique identifier.

        Returns:
            Flattened list of :class:`~kavachos.types.Permission` objects.
        """
        data = await self._http.request(
            "GET",
            f"/delegations/{quote(agent_id)}/permissions",
        )
        return [Permission.from_dict(p) for p in data]


class SyncDelegationResource:
    """Sync delegation chain operations."""

    def __init__(self, transport: SyncTransport) -> None:
        self._http = transport

    def create(self, input: DelegateInput) -> DelegationChain:
        """Create a delegation from one agent to another."""
        data = self._http.request("POST", "/delegations", json=input.to_dict())
        return DelegationChain.from_dict(data)

    def list_chains(self, agent_id: str) -> List[DelegationChain]:
        """List all delegation chains where the agent is a participant."""
        data = self._http.request("GET", f"/delegations/{quote(agent_id)}")
        return [DelegationChain.from_dict(d) for d in data]

    def revoke(self, delegation_id: str) -> None:
        """Revoke a specific delegation chain."""
        self._http.request("DELETE", f"/delegations/{quote(delegation_id)}")

    def get_effective_permissions(self, agent_id: str) -> List[Permission]:
        """Get the effective permissions for an agent, merging own and delegated."""
        data = self._http.request(
            "GET",
            f"/delegations/{quote(agent_id)}/permissions",
        )
        return [Permission.from_dict(p) for p in data]

File: sdks/go/delegation.go
package kavachos

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
)

// DelegationResource manages delegation chains between agents.
type DelegationResource struct {
	tp *transport
}

// Create creates a new delegation from one agent to another.
// The target agent receives a subset of the granting agent's permissions,
// subject to the depth limit.
func (r *DelegationResource) Create(ctx context.Context, input DelegateInput) (*DelegationChain, error) {
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "POST", "/delegations", input, nil, &raw); err != nil {
		return nil, err
	}
	var chain DelegationChain
	if err := json.Unmarshal(unwrapAPIData(raw), &chain); err != nil {
		return nil, fmt.Errorf("kavachos: decode delegation chain: %w", err)
	}
	return &chain, nil
}

// ListChains returns all delegation chains where the given agent is a participant.
func (r *DelegationResource) ListChains(ctx context.Context, agentID string) ([]DelegationChain, error) {
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "GET", "/delegations/"+url.PathEscape(agentID), nil, nil, &raw); err != nil {
		return nil, err
	}
	var chains []DelegationChain
	if err := json.Unmarshal(unwrapAPIData(raw), &chains); err != nil {
		return nil, fmt.Errorf("kavachos: decode delegation chains: %w", err)
	}
	return chains, nil
}

// Revoke deletes a delegation chain by its ID.
func (r *DelegationResource) Revoke(ctx context.Context, delegationID string) error {
	return r.tp.doJSON(ctx, "DELETE", "/delegations/"+url.PathEscape(delegationID), nil, nil, nil)
}

// GetEffectivePermissions returns the merged permissions for an agent, including
// all delegated grants.
func (r *DelegationResource) GetEffectivePermissions(ctx context.Context, agentID string) ([]Permission, error) {
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "GET", "/delegations/"+url.PathEscape(agentID)+"/permissions", nil, nil, &raw); err != nil {
		return nil, err
	}
	var perms []Permission
	if err := json.Unmarshal(unwrapAPIData(raw), &perms); err != nil {
		return nil, fmt.Errorf("kavachos: decode permissions: %w", err)
	}
	return perms, nil
}
